src/Analyzer/Passes/FuseFunctionsPass.cpp

```diff
@@ -0,0 +1,135 @@
+#include <Analyzer/Passes/FuseFunctionsPass.h>
+
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <Functions/FunctionFactory.h>
+
+#include <AggregateFunctions/IAggregateFunction.h>
+
+#include <Analyzer/InDepthQueryTreeVisitor.h>
+#include <Analyzer/FunctionNode.h>
+#include <Analyzer/ConstantNode.h>
+
+#include <DataTypes/DataTypesNumber.h>
+#include <DataTypes/DataTypeTuple.h>
+
+namespace DB
+{
+
+namespace
+{
+
+class FuseFunctionsMatcher : public InDepthQueryTreeVisitor<FuseFunctionsMatcher>
+{
+public:
+
+    static bool matchFunctionName(const String & name)
+    {
+        return name == "count" || name == "sum" || name == "avg";
+    }
+
+    void visitImpl(QueryTreeNodePtr & node)
+    {
+        auto * function_node = node->as<FunctionNode>();
+        if (!function_node || !function_node->isAggregateFunction() || !matchFunctionName(function_node->getFunctionName()))
+            return;
+
+        auto argument_hash = function_node->getArgumentsNode()->getTreeHash();
+        mapping[argument_hash].push_back(&node);
+    }
+
+    struct QueryTreeHashForMap
+    {
+        size_t operator()(const IQueryTreeNode::Hash & hash) const { return hash.first ^ hash.second; }
+    };
+
+    /// argument -> list of sum/count/avg functions with this argument
+    std::unordered_map<IQueryTreeNode::Hash, std::vector<QueryTreeNodePtr *>, QueryTreeHashForMap> mapping;
+};
+
+template <typename... Args>
+QueryTreeNodePtr createResolvedFunction(ContextPtr context, const String & name, DataTypePtr result_type, Args &&... args)
+{
+    auto function_node = std::make_shared<FunctionNode>(name);
+    auto function = FunctionFactory::instance().get(name, context);
+    function_node->resolveAsFunction(function, result_type);
+    function_node->getArguments().getNodes() = { std::forward<Args>(args)... };
+    return function_node;
+}
+
+QueryTreeNodePtr createTupleElement(ContextPtr context, DataTypePtr result_type, QueryTreeNodePtr argument, UInt64 idx)
+{
+    return createResolvedFunction(context, "tupleElement", result_type, argument, std::make_shared<ConstantNode>(idx));
+}
+
+QueryTreeNodePtr createSumCount(const FunctionNode & function_node, ContextPtr context)
```

Comment:
This function does 2 things, it create `sumCount` function and it replace `count`, `avg`, `sum` `function_node` with `tupleElement` with single `sumCount` function as argument.
I would prefer that `sumCount` function will be created once outside of this function. And this function will replace `function_node` with `tupleElement` depending on its name.